### DISCUSSION TITLE
Support for client routing by explicit Links

### DISCUSSION THREAD

**Discussion by jahredhope** [vikejs/vike#728] [Discussion] [answered]
As I understand it VPS will inject an event listener to monitor for any `<a />`  tags being clicked when Client Routing is enabled.

I'm familiar with React Router and other alternatives that require explicit `<Link />` components, rather than listening on all links. I generally try and avoid global behaviour when possible, and so I wanted to move to React Router's pattern.

I've attempted to re-create this behaviour by disabling `clientRouting` and creating my own `Link` component that calls `navigate`. However, there doesn't appear to be a way to opt into the client router in general, whilst opting out of the global `onLinkClick` behaviour specifically.

So my question is, can I have my cake and eat it to? 🍰  Can I use `clientRouting`, enjoy all the benefits of VPS's routing solution, whilst allowing me to handle the Link/navigate behaviour myself without global injection?

If not, I'd also be happy to be told why what I'm asking for is a bad idea, I'm very early on in a project so very happy to pivot.

Any advice is appreciated, thank-you for the awesome project.

**Comment by brillout** [comment]
I'm not aware and I don't see any drawbacks with VPS's approach. Hence, I'd recommend to stick to the beaten path, i.e. to how VPS intercepts clicks. Let me know if you run into any issue with that.

**Reply by dajinchu** [reply]
Hi, I am running into this as well. I'm migrating a large codebase that already intercepts all link clicks and has its own conventions. For backwards compatibility, I'd like to keep those conventions and just adapt our existing client router to call VPS' `navigate`.

As I understand it, I could just add `rel="external"` to all links, but it would be quite messy - I would love to just turn off VPS's `onLinkClick` behavior so I can use my own.

Thanks!

**Reply by brillout** [reply]
@dajinchu Use Server Routing instead of Client Routing.

**Reply by dajinchu** [reply]
Thanks @brillout - I still would like to call VPS' `navigate` to take advantage of all the infrastructure VPS has for handling pageContext and not triggering SSR on navigation. Basically I just need to override the link interception convention but keep the rest of Client Routing. Hope that makes sense! Thanks.

**Reply by brillout** [reply]
It does make sense. However, as you can imagine, the to-do list of VPS isn't small and it seems like you're the only user that would benefit from such feature. That said, if [sponsoring](https://github.com/sponsors/brillout) is an option for your company, then I'd be up to further discuss how we can solve your issue.

**Reply by jahredhope** [reply]
I very much appreciate the prioritization angle, but just wanted to confirm that I'm in the same boat. That's still only 2 though, and I was only validating the product, not an active consumer yet.

Are you open to pull-requests, if you think the ability to configure it is valuable but not a priority?

**Reply by brillout** [reply]
Yes, AFAICT it's doable in a PR. Basically adding a global option to make it opt-in instead of opt-out e.g. with `rel="internal"` although I'm not if it would be the right naming (I'm unfamiliar with the `rel="external"` naming).

**Reply by jahredhope** [reply]
Thanks.

> I'm not aware and I don't see any drawbacks with VPS's approach. Hence, I'd recommend to stick to the beaten path, i.e. to how VPS intercepts clicks. Let me know if you run into any issue with that.

Sorry, I realised I never replied to your original reply. I thought about this further and I'm still not overly comfortable with the global listener but I did trial it for my validation project and had no further issues (not that I'd expect them at smaller scale).

**Beaten path**
I'm not sure I'd personally characterise VPS's solution as the beaten path. One of the reasons I'm more comfortable with approaches like react-router's `<Link>` component is that it's a well-known common-place pattern. This may be a bias due to my React background though, and the global listener approach may be more common elsewhere.

**Drawbacks**
As for the drawbacks, I'd say they are broadly the same as any global behaviour. As a project scales it can become difficult to understand the impact of changes, and debug effects, when they aren't explicit.

For example, an engineer may have a shared component that works well in other projects, but when installed into a VPS project, finds the hack they have applied in the component doesn't work. As the engineer has only used an `<a />` there is nothing in there code that tells them to look at VPS's client-side routing documentation. Where-as if they see an imported `navigate` function they can track the import, read the docs, and realise their mistake.

**Reply by brillout** [reply]
I agree, predictability is an argument for making it opt-in. On the other hand, making it that it "just works" is an argument for opt-out. I'm inclined to think it's a trade-off situation where one way isn't necessarily superior to the other.

That said, if the global listener causes *concrete* problems, then that's a strong argument.

So far, I'd still go for opt-out. At least keeping it as VPS's default behavior (let alone because it would be a breaking change otherwise).

**Reply by brillout** [reply]
@jahredhope Would a new config `disableLinkClickInterceptor` with a new attribute `<a href="/product/42" data-vike-link />` work for you?

If you set `data-vike-link` then any link click is intercepted and VPS's client routing kicks in (i.e. the default behavior today). If you don't set `data-vike-link` then the link isn't intercepted and VPS lets the browser navigate to the new URL (the HTML of `href` is fetched, i.e. same behavior as `rel="external"`).

**Reply by jahredhope** [reply]
Yes. That makes sense and would support my use-case.

**Reply by sophiabarnes87** [reply]
Nice solution

**Reply by Irrelon** [reply]
That solution would be perfect. We have a concrete reason for prefering an opt-in. We have a catch-all in the +route.ts (export default "*") and work out what page to render based on a number of factors (query / search params etc). We are using dynamic / data driven config to allow pages to be configured and responded to at runtime.

Presently, other links on the page that are NOT our application's links but are still being handled by the vike router and as such, result in odd behaviour. We have already abstracted all our application anchor tags so adding an opt-in via a data attribute to any links our application renders will be easy.

I've tried multiple ways to get a custom +route.ts hook function to result in a behaviour similar to rel="external" but cannot find a way to do it. Our requirement is basically:

1) User clicks link
2) +route.ts function is called and determines if this route should respond via a `throw render("some actual page")`
3) If the route should NOT respond with a page from our application, it means the user is navigating to a page on the server that our application is not in control of, and we want the browser to hit the server and request the page

Either an opt-in for links so client routing doesn't kick in unless we specifically and explicitly declare the link as part of our app, or a way to indicate to the +route.ts output that the server should be handling the request, or preferably both would be great.

Something in +route.ts like `throw dontIntercept("some page path")` where the browser handles the link as if no event handler existed would be great!

**Reply by brillout** [reply]
@Irrelon Have you seen https://github.com/vikejs/vike/issues/918#issuecomment-1575633459? It seems like it would unblock you.

**Reply by Irrelon** [reply]
@brillout Thanks, I think this will work. Will test...

**Reply by brillout** [reply]
@Irrelon :+1: Let me know how it goes. FYI I agree with everything you said, e.g. it would be nice for Route Functions to be able to opt-out. Also, in case your company is up for it, we're looking for sponsors (#1350).

**Reply by Irrelon** [reply]
@brillout It works! Thanks. Regarding sponsorship, this is a possibility. I'll have a chat with the team :)

**Reply by brillout** [reply]
Great, I'm glad it works. Ok neat let me know how it goes!

**Comment by brillout** [comment, marked as answer]
Closing in favor of https://github.com/brillout/vite-plugin-ssr/issues/918.
